User's Templates (#17)
scramjetorg/scramjet-cloud-docs: README.md

# **Scramjet Transform Hub - Quick Start introduction**

> :bulb: **Note**: This repository contains introductory documentation and code samples for Scramjet Transform Hub. 

> * Developers looking for source code repository should visit [https://github.com/scramjetorg/transform-hub](https://github.com/scramjetorg/transform-hub). 

Reviewer: ```suggestion
* Developers looking for source code repository should visit the following link [https://github.com/scramjetorg/transform-hub](https://github.com/scramjetorg/transform-hub).
```

Our engine outputs can be managed in several ways: 

* **File** - you can save your output to local or remote file
* **STDOUT** - output can be directed to system STDOUT (STDERR is supported as well)
* **API** - output can be consumed from our  STH REST API
* **URL Request** - you can write your app in a way to request URL, webhook etc
* **Stream** - output can be streamed to a particular destination
* you can mix multiple actions together: you can both send data to remote system/URL and save it locally.

## **3. Install Scramjet Transform Hub**
### **3.1 Prepare environment**
In order to install Scramjet Transform Hub, please follow these 3 steps:
1. Get Linux machine (local UNIX/Linux OS, cloud VM etc)
2. Install Docker on this Linux machine ([official Docker instructions are here](https://docs.docker.com/get-docker/)) 
3. Install npm on this machine ([official instructions are here](https://nodejs.org/)). Currently we recommend Node.js version 16.x LTS.
### **3.2 Install STH**

> Open one Linux terminal window and issue following commands:

Reviewer: ```suggestion
Open a new terminal window and execute the following commands:
```

> **1. Install Scramjet Transform Hub and  STH CLI**

Reviewer: ```suggestion
**1. Install Scramjet Transform Hub and STH CLI**
```

```
scramjet-transform-hub
```
## **4. Run your first sequence**
### **4.1 Review the package**

> :bulb: **Note** that all commands here are executed from the root of this repository

> We have prepared for you a simple "hello-snowman" sequence that you can use. This sequence is available in the directory `samples/hello-snowman` in this repository.

Reviewer 2: There are some changes in the file that are worth changing. 

In 4.2 Prepare and send sequence package we have:
```
a) pack directory hello-snowman into archive hello-sequence.tar.gz
```
It should be changed to:

```
Make an archive file from the directory hello-snowman.
```

We can zip, compress, make an archive, etc.

This particular application is written in plain JavaScript to simplify this example. However, you can also write your sequences in TypeScript and build them before packaging and sending sequence to STH.

> In the template's [readme](templates/README.md) you will find a more specific descriptions of the particular file's content.

Reviewer: ```suggestion
In the template's [readme](templates/README.md) you will find a more detail description of a particular file's content.
```

> There is no need to change anything in our `hello-snowman` sequence for a first run. Let's move to the next step.

Reviewer 2: ```suggestion
There is no output shown in the terminal, but you can verify with ls that hello-sequence.tar.gz package is inside the samples directory.
```

### **4.2 Prepare and send sequence package**

Our "sequence" apps need to be packaged before sending to Transform Hub. Package is a simple TAR archive and our STH CLI has a special command to pack an app directory into a sequence tarball.

> > :bulb: **Note:** any time, you can display STH CLI help by issuing terminal command `si help` (for general help) or `si <command> help` for specific command (ie. `si sequence help`)

Reviewer: ```suggestion
> :bulb: **Note:** any time, you can display STH CLI help by issuing terminal command `si help` (for general help) or `si <command> help` for a specific command (ie. `si sequence help`)
```

> Please open new terminal window (and keep the first one open with STH running). Then issue following commands in the root directory of this repository:

Reviewer: ```suggestion
Please open a new terminal window (and do NOT close the first one open with STH running). Then issue the following commands in the root directory of this repository:
```

  instanceURL: 'instance/e70222d1-acfc-4e00-b046-4a3a9481c53b'
}
```

> Sequence is an app template. Once run, it will become live as a new instance. Instance also receives its own ID (GUID). In this case instance ID is:

Reviewer: ```suggestion
Sequence is an app template. Once it is up and running, it will become a new instance. Instance also receives its own ID (GUID). In this case instance ID is:
```

`_id: 'e70222d1-acfc-4e00-b046-4a3a9481c53b'`

> Of course, sequences can be run multiple times. Each run will create separate instance with a distinct instance ID.

Reviewer: ```suggestion
Of course, sequences can be run multiple times. Each run will create a separate instance with a distinct instance ID.
```

- in the other case (temperature above 0 degrees), sequence will return message `Snowman is melting! :(`

> To run this app, please execute this command from the root of our directory `node ./tools/stream-gen-tool/stream-gen.js <instance_id>`. In our case this would look like this:

Reviewer: ```suggestion
To run this app, please execute the following command from the root of our directory `node ./tools/stream-gen-tool/stream-gen.js <instance_id>`. In our case this would look like this:
```

2021-08-09T04:29:43.801Z log (object:Runner) Input message <Buffer 2d 33 36>
...
```

> > Well done! :clap::clap::clap: You have run your first Scramjet Transform Hub sequence! 

Reviewer: Ray suggests "[congratulations](https://www.youtube.com/watch?v=SC4xMk98Pdc&ab_channel=PostMaloneVEVO)" instead. His reasoning for this is the users just followed a series of steps. I'm leaving it to @a-tylenda to decide.
```
> Congratulations! :clap::clap::clap: You have run your first Scramjet Transform Hub sequence! 
```

Author: Congratulations it is 🤝 I will change it in my IDE and commit later. Thanks Ray!

## **5. Where to go next**
Here you can find more resources related to Scramjet Transform Hub:

> - [Check out more samples](samples) :books: - we have prepared some ready-to-use apps, which you can either use as a starting point for your own sequences or simply run them just to see what they do, and how STH works ith them.

Reviewer: ```suggestion
- [Check out more samples](samples) :books: - we have prepared some ready-to-use apps, which you can either use as a starting point for creating your own sequences or simply run them just to see what they do, and how the STH works with them.
```

> - [Start from our app templates](templates) :file_folder: - almost blank file structure (package) and use instructions, ready to be used as a starting point for your own sequences. This is the simplest base we can provide for you to start with.

Reviewer: ```suggestion
- [Start from our app templates](templates) :file_folder: - almost  a blank file structure (package) and usage instructions, ready to be used as a starting point for building your own sequences. This is the simplest base we can provide for you to start with.
```
